x/merkledb/cache.go

```diff
@@ -71,11 +71,16 @@ func (c *onEvictCache[K, V]) Flush() error {
 		c.lock.Unlock()
 	}()
 
+	// Note that we can't use [c.fifo]'s iterator because [c.onEviction]
+	// modifies [c.fifo], which violates the iterator's invariant.
 	var errs wrappers.Errs
-	iter := c.fifo.NewIterator()
-	for iter.Next() {
-		errs.Add(c.onEviction(iter.Value()))
```

Comment:
Note: previously we didn't remove the value - which meant that values that were iterated over here may have been written twice (in addition to having the iterator terminate early incorrectly)